File: Companies/GoldmanSachs/Permutations.java
package Companies.GoldmanSachs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @see Companies.Bloomberg.CombinationSum
 * @see Subsets
 */
public class Permutations {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> re = new ArrayList<>();
        dfs(re, new ArrayList<>(), nums, new boolean[nums.length]);
        return re;
    }

    private void dfs(List<List<Integer>> re, List<Integer> l, int[] nums, boolean[] used) {
        if (l.size() == nums.length) {
            re.add(new ArrayList<>(l));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i] || i > 0 && nums[i] == nums[i-1] && !used[i-1]) { // contain duplicates
                continue;
            }
            used[i] = true;
            l.add(nums[i]);
            dfs(re, l, nums, used);
            l.remove(l.size()-1);
            used[i] = false;
        }
    }
}

File: Topics/Design/al/Permutations.cpp
// Input: nums = [1,2,3]
// Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> result;
        vector<int> path;
        vector<bool> used(nums.size(), false);

        dfs(nums, used, path, result);
        return result;
    }

private:
    void dfs(const vector<int>& nums, vector<bool>& used, vector<int>& path, vector<vector<int>>& result) {
        if (path.size() == nums.size()) {
            result.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); ++i) {
            if (used[i]) {
                continue;
            }

            // if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) { // if contain duplicates
            //     continue;
            // }

            used[i] = true;
            path.push_back(nums[i]);

            dfs(nums, used, path, result);

            path.pop_back();
            used[i] = false;
        }
    }
};
